rs/ledger_suite/tests/sm-tests/src/lib.rs

```diff
@@ -2608,6 +2646,339 @@ pub fn test_downgrade_from_incompatible_version<T>(
+}
+
+pub fn test_migration_resumes_from_frozen<T>(
+    ledger_wasm_mainnet: Vec<u8>,
+    ledger_wasm_current_lowinstructionlimits: Vec<u8>,
+    encode_init_args: fn(InitArgs) -> T,
+) where
+    T: CandidType,
+{
+    let account = Account::from(PrincipalId::new_user_test_id(1).0);
+    let initial_balances = vec![(account, 100_000_000u64)];
+
+    let subnet_config = SubnetConfig::new(SubnetType::Application);
+    let env = StateMachine::new_with_config(StateMachineConfig::new(
+        subnet_config.clone(),
+        HypervisorConfig::default(),
+    ));
+
+    let args = encode_init_args(init_args(initial_balances));
+    let args = Encode!(&args).unwrap();
+    let canister_id = env
+        .install_canister_with_cycles(
+            ledger_wasm_mainnet,
+            args,
+            None,
+            Cycles::new(1_000_000_000_000),
+        )
+        .unwrap();
+
+    const APPROVE_AMOUNT: u64 = 150_000;
+    const NUM_APPROVALS: u64 = 20;
+
+    let send_approvals = || {
+        for i in 2..2 + NUM_APPROVALS {
+            let spender = Account::from(PrincipalId::new_user_test_id(i).0);
+            let approve_args = default_approve_args(spender, APPROVE_AMOUNT);
+            send_approval(&env, canister_id, account.owner, &approve_args)
+                .expect("approval failed");
+        }
+    };
+
+    send_approvals();
+
+    let check_approvals = || {
+        for i in 2..2 + NUM_APPROVALS {
+            let allowance = get_allowance(
+                &env,
+                canister_id,
+                account,
+                Account::from(PrincipalId::new_user_test_id(i).0),
+            );
+            assert_eq!(allowance.allowance, Nat::from(APPROVE_AMOUNT));
+        }
+    };
+
+    check_approvals();
+
+    env.upgrade_canister(
+        canister_id,
+        ledger_wasm_current_lowinstructionlimits,
+        Encode!(&LedgerArgument::Upgrade(None)).unwrap(),
+    )
+    .unwrap();
+
+    let is_ledger_ready = || {
+        Decode!(
+            &env.query(canister_id, "is_ledger_ready", Encode!().unwrap())
+                .expect("failed to call is_ledger_ready")
+                .bytes(),
+            bool
+        )
+        .expect("failed to decode is_ledger_ready response")
+    };
+    assert!(!is_ledger_ready());
+
+    let freeze = |env: &StateMachine, canister_id: CanisterId| {
+        let args = CanisterSettingsArgsBuilder::new()
+            .with_freezing_threshold(1 << 62)
+            .build();
+        let result = env.update_settings(&canister_id, args);
+        assert_matches!(result, Ok(_));
+    };
+    let unfreeze = |env: &StateMachine, canister_id: CanisterId| {
+        let args = CanisterSettingsArgsBuilder::new()
+            .with_freezing_threshold(0)
+            .build();
+        let result = env.update_settings(&canister_id, args);
+        assert_matches!(result, Ok(_));
+    };
+
+    freeze(&env, canister_id);
+    env.advance_time(Duration::from_secs(1000));
+    env.tick();
```

Comment:
I'd tick multiple times, e.g., 10x, to make sure the timer was attempted to be scheduled.

Reply:
> Is test_migration_resumes_from_frozen what you had in mind? In the test I upgrade, test_not_ready, freeze, wait 1000s, unfreeze, test_not_ready, wait_until_ready.

Yes, modulo my above comment.

Reply:
Done!